File: LinkedList.java
public class LinkedList {
    public static void main(String[] args) {
        LinkedListExample ll = new LinkedListExample();

        ll.add("1");
        ll.add("2");
        ll.add("3");
        ll.add("4");
        ll.add("FU");

        System.out.println("Whole list:" + ll);
    }
}

class LinkedListExample {
    private Node head;
    private static int counter;

    public void add(Object data) {

        //If first element in the node
        if (head ==null) {
            System.out.println("Create Head");
            head = new Node();
        }

        //Create new node
        Node currentData = new Node(data);
        Node currentNode = head;

        // Check for NPE
        if (currentNode !=null) {
                while (currentNode.getNext() !=null) {
                currentNode = currentNode.getNext();
            }

            currentNode.setNext(currentData);

        }

        incrementCounter();
    }

    private static int getCounter() {
		return counter;
	}

	private static void incrementCounter() {
		counter++;
	}

	private void decrementCounter() {
		counter--;
	}

    public String toString() {
		String output = "";

		if (head != null) {
			Node currentNode = head.getNext();
			while (currentNode != null) {
				output += "[" + currentNode.getData().toString() + "]";
				currentNode = currentNode.getNext();
			}

		}
		return output;
	}

    private class Node {
        Node next;
        Object data;

        public Node() {
            next = null;
            data = null;
        }

        public Node(Object value) {
            next = null;
            data = value;
        }

        //2nd constructor
        public Node(Object value, Node nextValue) {
            next = nextValue;
            data = value;
        }

        public Object getData() {
            return data;
        }

        public void setData(Object value) {
            data = value;
        }

        public Node getNext() {
            return next;
        }

        public void setNext(Node nextValue) {
            next = nextValue;
        }
    }

}

File: linkedList.js
// Create a LinkedList - set head to null on intialization
function LinkedList(){
    this.head = null;
}
// Push function to add a node
LinkedList.prototype.push = function (value) {
    var node = {
            data: value,
            next: null
        },
        currentNode;

    //Make this node the head if it is null
    if (!this.head) {
        this.head = node;
    }
    else {
        currentNode = this.head;
        // while there is a next element set current to the next one.
        while (currentNode.next) {
            currentNode = currentNode.next
        }
        // Add the node to next at the end of the LL
        currentNode.next = node;
    }
}

LinkedList.prototype.pop = function(value){
    var currentNode = this.head;

    // check if pop is for head element
    if(currentNode.data == value){
       this.head = currentNode.next;
    }
    else{
      var previous = currentNode;

      while(currentNode.data){
        //case-3
        console.log(value)
        console.log("Yes",currentNode.data)
        if(currentNode.data == value){
        console.log("Fuck you.")
          previous.next = currentNode.next;
          break;
        }
        previous = currentNode;
        currentNode = currentNode.next;
      }
      //case -2
      if(currentNode.data == value){
        previous.next == null;
      }
    }
}

console.log("Single Linked List example - in JS");
var singleLinkedList = new LinkedList();

//push nodes
singleLinkedList.push(555);
singleLinkedList.push("Hello");
singleLinkedList.push("Goodbye");

console.log(singleLinkedList.head);
console.log(singleLinkedList.head.next);
console.log(singleLinkedList.head.next.next);

console.log("Pop - Goodbye");
singleLinkedList.pop("Goodbye");
console.log(singleLinkedList.head);

console.log("Pop - Goodbye");
singleLinkedList.pop(555);
console.log(singleLinkedList.head);
